crates/ty_python_semantic/resources/mdtest/narrow/len.md

```diff
@@ -0,0 +1,128 @@
+# Narrowing for `len(..)` checks
+
+When `len(x)` is used in a boolean context, we can narrow the type of `x` based on whether `len(x)`
+is truthy (non-zero) or falsy (zero).
+
+We apply `~AlwaysFalsy` narrowing when ANY part of the type is narrowable (string/bytes literals,
+`LiteralString`, tuples). This removes types that are always falsy (like `Literal[""]`) while
+leaving non-narrowable types (like `str`, `list`) unchanged.
+
+## String literals
+
+The intersection with `~AlwaysFalsy` simplifies to just the non-empty literal.
+
+```py
+from typing import Literal
+
+def _(x: Literal["foo", ""]):
+    if len(x):
+        reveal_type(x)  # revealed: Literal["foo"]
+    else:
+        reveal_type(x)  # revealed: Literal[""]
+```
+
+## Bytes literals
+
+```py
+from typing import Literal
+
+def _(x: Literal[b"foo", b""]):
+    if len(x):
+        reveal_type(x)  # revealed: Literal[b"foo"]
+    else:
+        reveal_type(x)  # revealed: Literal[b""]
+```
+
+## LiteralString
+
+```toml
+[environment]
+python-version = "3.11"
+```
+
+```py
+from typing import LiteralString
+
+def _(x: LiteralString):
+    if len(x):
+        reveal_type(x)  # revealed: LiteralString & ~Literal[""]
+    else:
+        reveal_type(x)  # revealed: Literal[""]
+```
+
+## Tuples
+
+```py
+def _(x: tuple[int, ...]):
+    if len(x):
+        reveal_type(x)  # revealed: tuple[int, ...] & ~AlwaysFalsy
+    else:
+        reveal_type(x)  # revealed: tuple[int, ...] & ~AlwaysTruthy
+```
+
+## Unions of narrowable types
+
+```py
+from typing import Literal
+
+def _(x: Literal["foo", ""] | tuple[int, ...]):
+    if len(x):
+        reveal_type(x)  # revealed: Literal["foo"] | (tuple[int, ...] & ~AlwaysFalsy)
+    else:
+        reveal_type(x)  # revealed: Literal[""] | (tuple[int, ...] & ~AlwaysTruthy)
+```
+
+## Types that are not narrowed
+
+For `str`, `list`, and other types where a subclass could have a `__bool__` that disagrees with
+`__len__`, we do not narrow:
+
+```py
+def not_narrowed_str(x: str):
+    if len(x):
+        # No narrowing because `str` could be subclassed with a custom `__bool__`
+        reveal_type(x)  # revealed: str
+
+def not_narrowed_list(x: list[int]):
+    if len(x):
+        # No narrowing because `list` could be subclassed with a custom `__bool__`
```

Comment:
and even if it couldn't: the `__bool__` of `list` can anyway give inconsistent results throughout the course of a function. E.g. if you have this, say we know for sure that `x` is _exactly_ `list[int]`, not a subclass thereof:

```py
def f(x: list[int]):
    if len(x):
        reveal_type(x)  # say we infer it as `list[int] & AlwaysTruthy` now
        x.clear()
        reveal_type(x)  # what's the type now? Still `list[int] & AlwaysTruthy`?
                        # how do we detect that the `.clear()` call invalidates the previous narrowing?

        do_something_with_x(x)  # does _this_ function call invalidate the narrowing?
                                # it's a black box to us: we don't know what happens inside it
```